v1.0.0 update

domhoward14 opened the pull request.

domhoward14 commented:
@anilsenay I addressed all your points below. Points 1,2, and 4 I think I understand and i'll push up another change tomorrow or saturday addressing those issues.

1. Yeah, I removed some content from the readme temporarily so that I could test and understand it better. Now that the plan is to merge it back into the original repo, I will add back the missing content. Most of the additions I made are placed in the tables, but I will double-check to ensure that I haven't missed anything important. I admit that I rushed through the README a little, so it's possible that I may have skipped some crucial details 😅.

2. Correct me if I am misunderstanding your concern, but I believe I understand the core of your issue here. In fact, I had already anticipated this problem and had a TODO for it because I knew it would be a major challenge for users lol. This led me to two key points.
   a. Firstly, creating another function solely responsible for setting the `Path` attribute won't scale well. It will lead to the old problem of inheritance scaling. We will have to create a `Path` function for every parameter type we support, and the same goes for every other location (e.g., `IntPath`, `StrPath`, `BoolPath`, `IntHeader`, `StrHeader`, `BoolHeader`, etc.). The solution I had in mind (which I just left as a TODO due to laziness lol) was to make the `Path` attribute a requirement rather than an option for any `Parameter` type. This approach will address the issue you described, and we won't need to support all the types, which would otherwise lead the number of new functions we had to support = (Parameter types * Parameter Locations).

    b. Secondly, this is related to a larger point, but also this one. I don't think the initial version of Swagno should be overly focused on validating user input. I believe we should iterate on our initial release and then potentially introduce an optional object or function for clients to use in the future that checks for issues like forgetting to set the `Path` attribute and other semantic problems, by having the function returning an error. However, for the first iteration, it would be simpler to let the user take responsibility for fixing most semantic issues.

3. Hmmm, I don't have an answer for this. I thought I didn't encounter this issue during my testing. Could you please provide me with the struct and the method you used to create the endpoint so that I can test it on my end? If I had to guess, I think you're right about the parsing of the structs using functions like `getExampleTag` potentially causing the problem.

4.  I really like this solution. I did have a similar idea, but it didn't include the composition aspect that you added. My solution would have required refactoring the generation function (and once again my laziness prevented that), but yours works much more elegantly, thanks to the composition of the `Model` attribute. I will test this on my end and get back to ya.

anilsenay commented:
About 1., I liked the refactor you did in README, thanks. If you have not enough time for adding missing content etc., just copy/paste from original 😄 I can do the final refactor on it before merging

About 2.a., I got your point and it would be better if we simplified it. But the reason why I seperated them for each function was that I thought we could manipulate them easier for each type. If you come with a better approach in your mind, it would be great. In this time, I will also think about it

About 2.b, Ye, validation would be great. Also I think we should add this in README for future improvement (https://github.com/go-swagno/swagno#implementation-status)

About 3., I think model is not imported here because I created a new struct type and it had the same issue:
```go
type Response2 struct {
	Id int `json:"id"`
}

func (s Response2) GetDescription() string {
	return "Test"
}
func (s Response2) GetReturnCode() string {
	return "500"
}

// ...
endpoint.WithSuccessfulReturns([]response.Info{Response2{}}),
// ...
``` 

![resim](https://github.com/go-swagno/swagno/assets/1047345/4157d11c-4d76-4ca0-9f3e-518213e26ed6)

About 4., Alright 👍

anilsenay commented on example/server.go:
```diff
@@ -0,0 +1,68 @@
+		),
+	}
+
+	// TODO make support for popular http server libs so that the creation of endpoints and handlers can happen in one funciton
```
How can we achieve this? Because all web frameworks accept different types of `HandlerFunc`. For example: `gin.HandlerFunc`, `func(c *fiber.Ctx) error` etc. Thats why I created different packages for [gin](https://github.com/go-swagno/swagno-gin) and [fiber](https://github.com/go-swagno/swagno-fiber).

If we can not achieve this at the moment, we should add fiber and gin examples additionally I guess 😄

anilsenay replied:
I know fiber has adaptor for accepting `http.HandlerFunc` but I am not sure all frameworks also have adaptor-like feature

domhoward14 replied:
Yeah, I was going to come back to this one. I want to research this one a little more. For now since we're merging back to the original repo I'll revert the gin and fiber examples like you suggested.

anilsenay commented on go.mod:
```diff
@@ -1,40 +1,14 @@
-module github.com/go-swagno/swagno
+module github.com/domhoward14/swagno
```
you should change it back 😄  This change also effect all of your imports in all files. You should also replace them as `github.com/go-swagno/swagno`  😅

anilsenay commented on README.md:
```diff
@@ -1,20 +1,80 @@
 
-Swagno is an approach to create Swagger Documentation 2.0 without any **annotation**, **exported file** or any **command** to run.
-You can declare your documentation details in code and get a json string to serve with a handler.
+Swagno offers a streamlined approach to create Swagger Documentation 2.0. With Swagno, you can declare documentation details directly in code, eliminating the need for annotations, exported files, or commands. This approach simplifies the process of generating a JSON string to serve with a handler.
```
This is nice description but I like swagno's punchline: 
``` `no` annotation, `no` exported file,  `no` command to run ``` 😄
It describes the meaning of "swag`no`". Can you adapt this punchline in this description somehow 😄 Its not `must` btw you can skip this comment if you don't want to add.

anilsenay commented on README.md:
```diff
@@ -1,20 +1,80 @@
 
 This project inspired by [Swaggo](https://github.com/swaggo/swag). Swaggo, uses annotations, exports files and needs to run by command. If you don't like this way, [Swag**no**](https://github.com/go-swagno/swagno) appears as a good alternative.
 
+This project was then forked from [go-swagno/swagno](https://github.com/go-swagno/swagno) with the goal of being more idiomatic and  user friendly. While trying to achieve these goals a lot of breaking changes were made and hence why this repo is separate as opposed to being merged into the original.
```
We can clear this from L:11 to L:73 but we can add this in `release notes`

domhoward14 replied:
good idea. I translated these into release notes. Will upload once merge and release is ready.

anilsenay commented on README.md:
```diff
@@ -27,500 +87,318 @@ This project inspired by [Swaggo](https://github.com/swaggo/swag). Swaggo, uses
   - [Endpoints (API)](#endpoints-api)
     - [Arguments](#arguments)
 - [Contribution](#contribution)
+- [Examples](example/models/)
```
Why not just /example ?

anilsenay commented on README.md:
```diff
@@ -27,500 +87,318 @@ This project inspired by [Swaggo](https://github.com/swaggo/swag). Swaggo, uses
-  EndPoint(GET, "/product/{id}", "product", Params(IntParam("id", true, "")), nil, models.Product{}, models.ErrorResponse{}, "", nil),
-  EndPoint(GET, "/product/{id}/detail", "product", Params(IntParam("id", true, "")), nil, models.Product{}, models.ErrorResponse{}, "", nil),
-  EndPoint(POST, "/product", "product", Params(), models.ProductPost{}, models.Product{}, models.ErrorResponse{}, "", nil),
+go get github.com/domhoward14/swagno
```
You will probably revert module names back but I am commenting for just a reminder

anilsenay commented on README.md:
````diff
@@ -27,500 +87,318 @@ This project inspired by [Swaggo](https://github.com/swaggo/swag). Swaggo, uses
+ http.ListenAndServe(":8080", nil)
 ```
 
-## Supported Web Frameworks
````
We can add this section back by refactoring examples 😄

anilsenay commented:
> @anilsenay lol nah this is good! I appreciate the thoroughness 😀 . Like I said yesterday I'm going to start working on some of the feedback today. All of the refactoring involving adding stuff back in I should be able get done pretty quickly, but just give me a a couple of days for the code change updates 👍

Of course, take your time. In this period, I will work on supporting `map` and `slice` types in response. Maybe I can make it before the release

domhoward14 commented on swagger.go:
```diff
@@ -4,50 +4,55 @@ import (
+// Contact represents the contact information for the API.
 // https://swagger.io/specification/v2/#contact-object
-type swaggerContact struct {
+type contact struct {
```
missed that linter check. I've removed it.

domhoward14 replied:
i also added the missing fields for this struct that you pointed out

domhoward14 commented on components/parameter/parameter.go:
```diff
@@ -0,0 +1,491 @@
+
+// validate will check and correct certian parameter configurations that cause non-fatal OpenAPI warning/errors when rendering.
+// e.g. Options being set for a parameters `In == Path`, while `required == false` is a non-fatal syntactical error.
+func (p *Parameter) validate(l Location) {
```
Yup, I started to then I realized that to do this right it would be a bigger effort than i was ready to do now. Forgot to remove after quitting on that idea. It's gone now.

domhoward14 commented:
> > @anilsenay lol nah this is good! I appreciate the thoroughness 😀 . Like I said yesterday I'm going to start working on some of the feedback today. All of the refactoring involving adding stuff back in I should be able get done pretty quickly, but just give me a a couple of days for the code change updates 👍
> 
> Of course, take your time. In this period, I will work on supporting `map` and `slice` types in response. Maybe I can make it before the release

Awesome, sounds good 😁

domhoward14 commented:
@anilsenay I pushed updates for most of the easy refactoring stuff. For the 4 points you made earlier in the thread i'm going to save those for tomorrow. Thanks again for the thorough feedback!

anilsenay commented:
@domhoward14  I opened a PR in your repository, could you take a look it?

domhoward14 commented:
@anilsenay heads up, i'm going to double check and make sure that i addressed all your comments tomorrow and then push them up after. After I push it up i think we should continue to work on the other PR that you opened up so we can then merge that into this one and get this released!

domhoward14 commented:
@anilsenay Updates have been pushed. Let me know if there's anything i can do to assist with preparing for the release.

anilsenay reviewed (approved):
@domhoward14 Thanks for all your efford. I think its finally ready to merge. I will merge this pr and release a pre-release version as v1.0.0rc1. Tomorrow, I am planning to do a detailed developer experience test before v1 release.
